{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "String jsonText = new File('data/api-export/genes/genes-673.json').text\n",
    "def slurper = new groovy.json.JsonSlurper()\n",
    "json =slurper.parseText(jsonText)\n",
    "OutputCell.HIDDEN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "248f936f-ae66-4ef0-9d2c-dce879e4f1b7",
       "version_major": 2,
       "version_minor": 0
      },
      "method": "display_data"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "//  impact color map for lollies\n",
    "\n",
    "def VARIANT_IMPACT_COLORS = [\"deletion\": '#ED8B00', \"duplication\": '#333F48', \"frameshift\": '#A2AAAD',\n",
    "                                 \"fusion\"  : '#0085CA', \"indel\": '#9467bd', \"insertion\": '#43B02A', \"missense\": '#002D72',\n",
    "                                 \"none\"    : '#7f7f7f', \"nonsense\": '#bcbd22', \"splice site\": '#17becf', \"unknown\": '#D0D3D4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "// get maps of fields\n",
    "mapByImpact = json.variants.groupBy{it.impact}// factor\n",
    "mapByName = json.variants.groupBy{it.fullName}\n",
    "OutputCell.HIDDEN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "example of variant zero (0) object data ->\n",
      "\n",
      "fullname: BRAF S602F ==> position: 602\n",
      "x : 602\n",
      "y :27728\n",
      "factor :missense\n",
      "\n",
      "\n",
      "Full object first variable format: \n",
      "[id:27728, fullName:BRAF S602F, impact:missense, proteinEffect:no effect - predicted, geneVariantDescriptions:[[description:BRAF S602F lies within the protein kinase domain of the Braf protein (UniProt.org). S602F has not been biochemically characterized, but results in similar cell proliferation and viability levels to wild-type Braf in culture (PMID: 29533785), and therefore is predicted to have no effect on Braf protein function., references:[[id:11238, pubMedId:29533785, title:Systematic Functional Annotation of Somatic Mutations in Cancer., url:http://www.ncbi.nlm.nih.gov/pubmed/29533785], [id:295, pubMedId:null, title:null, url:www.uniprot.org]]]], position:602]\n",
      "\n",
      "\n",
      "mutation data object quality check :\n",
      "\n",
      "size : 263\n",
      "\n",
      "[id:27728, fullName:BRAF S602F, impact:missense, proteinEffect:no effect - predicted, geneVariantDescriptions:[[description:BRAF S602F lies within the protein kinase domain of the Braf protein (UniProt.org). S602F has not been biochemically characterized, but results in similar cell proliferation and viability levels to wild-type Braf in culture (PMID: 29533785), and therefore is predicted to have no effect on Braf protein function., references:[[id:11238, pubMedId:29533785, title:Systematic Functional Annotation of Somatic Mutations in Cancer., url:http://www.ncbi.nlm.nih.gov/pubmed/29533785], [id:295, pubMedId:null, title:null, url:www.uniprot.org]]]], position:602]\n"
     ]
    }
   ],
   "source": [
    "// build up genetic mutation data object by extracting relevant fields\n",
    "def impacts = mapByImpact.keySet() as List\n",
    "def impactCounts = mapByImpact.values().collect{it.size()}\n",
    "def names = mapByName.keySet() as List\n",
    "\n",
    "//println(\"names :\"+names)\n",
    "\n",
    "// regular expression matching to extract position\n",
    "def builder = new groovy.json.JsonBuilder(json)\n",
    "builder.content.variants.each{\n",
    "    def fullName = it.fullName\n",
    "    def group = fullName =~ /.* \\w([0-9]+)[^0-9]+/  \n",
    "    if(group){\n",
    "        def position = group[0][1]\n",
    "        it.position = position\n",
    "    }\n",
    "    else{\n",
    "        it.position = null\n",
    "    }\n",
    "    //println(\"position :\"+it.position+\"\\n\")\n",
    "} \n",
    "\n",
    "// test print (quality check)\n",
    "def map = builder.content.variants[0]\n",
    "println(\"example of variant zero (0) object data ->\\n\\n\"+\"fullname: ${map.fullName} ==> position: ${map.position}\")\n",
    "println(\"x : \"+map.position+\"\\ny :\"+map.id+\"\\nfactor :\"+map.impact )\n",
    "println(\"\\n\\nFull object first variable format: \\n\"+builder.content.variants[0])\n",
    "\n",
    "// updated JSON file\n",
    "new File('data/api-export/genes/genes-673-with-position.json').text = builder.toPrettyString()\n",
    "\n",
    "// build Genetic mutation data object\n",
    "def mutation_data = builder.content.variants\n",
    "\n",
    "// new JSON file of just variants\n",
    "//def slurper = new groovy.json.JsonSlurper()\n",
    "//json =slurper.parseText(mutation_data)\n",
    "//new File('data/api-export/genes/genes-673-variants.json').text = mutation_data\n",
    "\n",
    "\n",
    "beakerx.mutation_data = mutation_data\n",
    "\n",
    "println (\"\\n\\nmutation data object quality check :\\n\\n\"+\"size : \"+mutation_data.size()+\"\\n\\n\"+mutation_data[0])\n",
    "\n",
    "OutputCell.HIDDEN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "var mutation_data_default_settings = {\n",
       "    x: \"position\",         // mutation position\n",
       "    y: \"Protein_Change\",      // impact count\n",
       "    factor: \"impact\", // classify mutations by certain factor (optional)\n",
       "};\n",
       "\n",
       "beakerx.mutation_data_default_settings = mutation_data_default_settings"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%javascript\n",
    "\n",
    "var mutation_data_default_settings = {\n",
    "    x: \"position\",         // mutation position\n",
    "    y: \"fullname\",      // impact count\n",
    "    factor: \"impact\", // classify mutations by certain factor (optional)\n",
    "};\n",
    "\n",
    "beakerx.mutation_data_default_settings = mutation_data_default_settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "\n",
       "var d3 = require(['d3'], function (d3) {\n",
       "    console.log(\"hello d3\" + d3);\n",
       "});\n",
       "// generate g3-lollipop chart\n",
       "\n",
       "// create in \"g3lollipop\" div\n",
       "var g3 = require(['g3'], function (g3) {\n",
       "    console.log(\"hello g3\" + g3);\n",
       "    var lollipop = g3.Lollipop(\"g3lollipop\");\n",
       "\n",
       "    // add mutation data\n",
       "    lollipop.data.snvData = beakerx.mutation_data;\n",
       "    // mutation data format settings\n",
       "    lollipop.format.snvData = mutation_data_default_settings;\n",
       "\n",
       "    // Pfam domain data\n",
       "    lollipop.data.domainData = pfam_data;\n",
       "    // Pfam data format settings\n",
       "    lollipop.format.domainData = pfam_data_default_settings;\n",
       "\n",
       "    // set up more chart options ...\n",
       "    var plot_options = {\n",
       "        \"chartMargin\": {\n",
       "            \"left\": 40, \"right\": 40, \"top\": 30, \"bottom\": 25\n",
       "        },\n",
       "        \"titleFont\": \"normal 20px Sans\",\n",
       "        \"titleColor\": \"steelblue\",\n",
       "        \"titleAlignment\": \"middle\",\n",
       "        \"titleDy\": \"0.3em\"\n",
       "    };\n",
       "\n",
       "    // add chart options\n",
       "    lollipop.setOptions(plot_options);\n",
       "\n",
       "    lollipop.options.chartMargin = {\n",
       "    \"left\": 40, \"right\": 40, \"top\": 30, \"bottom\": 25};\n",
       "    lollipop.options.titleFont = \"normal 20px Sans\";\n",
       "    lollipop.options.titleColor = \"steelblue\";\n",
       "    lollipop.options.titleAlignment = \"middle\";\n",
       "    lollipop.options.titleDy = \"0.3em\";\n",
       "    \n",
       "    // draw lollipop\n",
       "    lollipop.draw();\n",
       "\n",
       "});"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%javascript\n",
    "\n",
    "\n",
    "var d3 = require(['d3'], function (d3) {\n",
    "    console.log(\"hello d3\" + d3);\n",
    "});\n",
    "// generate g3-lollipop chart\n",
    "\n",
    "// create in \"g3lollipop\" div\n",
    "var g3 = require(['g3'], function (g3) {\n",
    "    console.log(\"hello g3\" + g3);\n",
    "    var lollipop = g3.Lollipop(\"g3lollipop\");\n",
    "\n",
    "    // add mutation data\n",
    "    lollipop.data.snvData = beakerx.mutation_data;\n",
    "    // mutation data format settings\n",
    "    lollipop.format.snvData = mutation_data_default_settings;\n",
    "\n",
    "    // Pfam domain data\n",
    "    lollipop.data.domainData = pfam_data;\n",
    "    // Pfam data format settings\n",
    "    lollipop.format.domainData = pfam_data_default_settings;\n",
    "\n",
    "    // set up more chart options ...\n",
    "    var plot_options = {\n",
    "        \"chartMargin\": {\n",
    "            \"left\": 40, \"right\": 40, \"top\": 30, \"bottom\": 25\n",
    "        },\n",
    "        \"titleFont\": \"normal 20px Sans\",\n",
    "        \"titleColor\": \"steelblue\",\n",
    "        \"titleAlignment\": \"middle\",\n",
    "        \"titleDy\": \"0.3em\"\n",
    "    };\n",
    "\n",
    "    // add chart options\n",
    "    lollipop.setOptions(plot_options);\n",
    "\n",
    "    lollipop.options.chartMargin = {\n",
    "    \"left\": 40, \"right\": 40, \"top\": 30, \"bottom\": 25};\n",
    "    lollipop.options.titleFont = \"normal 20px Sans\";\n",
    "    lollipop.options.titleColor = \"steelblue\";\n",
    "    lollipop.options.titleAlignment = \"middle\";\n",
    "    lollipop.options.titleDy = \"0.3em\";\n",
    "    \n",
    "    // draw lollipop\n",
    "    lollipop.draw();\n",
    "\n",
    "});\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<html>\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "\n",
       "<head>\n",
       "    <meta charset=\"utf-8\">\n",
       "    <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n",
       "    <link rel=\"stylesheet\" href=\"https://s3-us-west-2.amazonaws.com/cdsjsd3/g3-viz/0.5.0/g3-lollipop.min.css\">\n",
       "    <link rel=\"stylesheet\" href=\"https://s3-us-west-2.amazonaws.com/cdsjsd3/css/gist.css\">\n",
       "</head>\n",
       "\n",
       "<body>\n",
       "    <div>\n",
       "        <div>\n",
       "            <span class=\"btn-group\">\n",
       "                <button id=\"save-as-png\">save as PNG</button>\n",
       "                <button id=\"save-as-svg\">save as SVG</button>\n",
       "            </span>\n",
       "        </div>\n",
       "\n",
       "        <div id=\"lollipop\"></div>\n",
       "    </div>\n",
       "\n",
       "    <script src=\"https://d3js.org/d3.v4.min.js\"></script>\n",
       "    <script src=\"https://s3-us-west-2.amazonaws.com/cdsjsd3/g3-viz/0.5.0/g3-lollipop.min.js\"></script>\n",
       "    <script>\n",
       "        // mutation data in TSV (tabular-separated value) format\n",
       "        var snvFile = \"TP53-msk_impact_2017-parsed.tsv\";\n",
       "\n",
       "        // pfam domain data\n",
       "        var domainFile = \"TP53_pfam.json\";\n",
       "\n",
       "        // mutation data format [to be modified]\n",
       "        var snvDataFormat = {\n",
       "            x: \"AA_Position\",\n",
       "            y: \"Protein_Change\",\n",
       "            factor: \"Mutation_Type\",\n",
       "        };\n",
       "\n",
       "        // pfam domain JSON information [to be modified]\n",
       "        var domainDataFormat = {\n",
       "            symbol: \"hgnc_symbol\",\n",
       "            name: \"protein_name\",\n",
       "            length: \"length\",\n",
       "            domainType: \"pfam\",\n",
       "            details: {\n",
       "                start: \"pfam_start\",\n",
       "                end: \"pfam_end\",\n",
       "                ac: \"pfam_ac\",\n",
       "                name: \"pfam_id\",\n",
       "            },\n",
       "        };\n",
       "\n",
       "        // read data in\n",
       "        var q = d3.queue();\n",
       "        q.defer(d3.tsv, snvFile);\n",
       "        q.defer(d3.json, domainFile);\n",
       "\n",
       "        q.await(function (error, snvData, domainData) {\n",
       "            // format data\n",
       "            snvData.forEach(function (d) {\n",
       "                d[snvDataFormat.x] = +d[snvDataFormat.x];\n",
       "            });\n",
       "\n",
       "            // new lollipop chart\n",
       "            var lollipop = g3.Lollipop(\"lollipop\");\n",
       "\n",
       "            // add data\n",
       "            lollipop.data.snvData = snvData;\n",
       "            lollipop.data.domainData = domainData;\n",
       "            // add data options\n",
       "            lollipop.format.snvData = snvDataFormat;\n",
       "            lollipop.format.domainData = domainDataFormat;\n",
       "\n",
       "            // draw with default settings\n",
       "            lollipop.draw();\n",
       "\n",
       "            // get chart div id\n",
       "            var chartID = lollipop.options.chartID;\n",
       "\n",
       "            // set default filename\n",
       "            var output_chart_filename = \"lollipop\";\n",
       "\n",
       "            // button actions: download chart\n",
       "            document.getElementById(\"save-as-png\").onclick = function (e) {\n",
       "                g3.output().toPNG(output_chart_filename, chartID);\n",
       "            };\n",
       "\n",
       "            document.getElementById(\"save-as-svg\").onclick = function (e) {\n",
       "                g3.output().toSVG(output_chart_filename, chartID);\n",
       "            };\n",
       "\n",
       "        });\n",
       "    </script>\n",
       "</body>\n",
       "\n",
       "</html></html>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%html\n",
    "\n",
    "<!DOCTYPE html>\n",
    "<html>\n",
    "\n",
    "<head>\n",
    "    <meta charset=\"utf-8\">\n",
    "    <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n",
    "    <link rel=\"stylesheet\" href=\"https://s3-us-west-2.amazonaws.com/cdsjsd3/g3-viz/0.5.0/g3-lollipop.min.css\">\n",
    "    <link rel=\"stylesheet\" href=\"https://s3-us-west-2.amazonaws.com/cdsjsd3/css/gist.css\">\n",
    "</head>\n",
    "\n",
    "<body>\n",
    "    <div>\n",
    "        <div>\n",
    "            <span class=\"btn-group\">\n",
    "                <button id=\"save-as-png\">save as PNG</button>\n",
    "                <button id=\"save-as-svg\">save as SVG</button>\n",
    "            </span>\n",
    "        </div>\n",
    "\n",
    "        <div id=\"lollipop\"></div>\n",
    "    </div>\n",
    "\n",
    "    <script src=\"https://d3js.org/d3.v4.min.js\"></script>\n",
    "    <script src=\"https://s3-us-west-2.amazonaws.com/cdsjsd3/g3-viz/0.5.0/g3-lollipop.min.js\"></script>\n",
    "    <script>\n",
    "        // mutation data in TSV (tabular-separated value) format\n",
    "        var snvFile = \"TP53-msk_impact_2017-parsed.tsv\";\n",
    "\n",
    "        // pfam domain data\n",
    "        var domainFile = \"TP53_pfam.json\";\n",
    "\n",
    "        // mutation data format [to be modified]\n",
    "        var snvDataFormat = {\n",
    "            x: \"AA_Position\",\n",
    "            y: \"Protein_Change\",\n",
    "            factor: \"Mutation_Type\",\n",
    "        };\n",
    "\n",
    "        // pfam domain JSON information [to be modified]\n",
    "        var domainDataFormat = {\n",
    "            symbol: \"hgnc_symbol\",\n",
    "            name: \"protein_name\",\n",
    "            length: \"length\",\n",
    "            domainType: \"pfam\",\n",
    "            details: {\n",
    "                start: \"pfam_start\",\n",
    "                end: \"pfam_end\",\n",
    "                ac: \"pfam_ac\",\n",
    "                name: \"pfam_id\",\n",
    "            },\n",
    "        };\n",
    "\n",
    "        // read data in\n",
    "        var q = d3.queue();\n",
    "        q.defer(d3.tsv, snvFile);\n",
    "        q.defer(d3.json, domainFile);\n",
    "\n",
    "        q.await(function (error, snvData, domainData) {\n",
    "            // format data\n",
    "            snvData.forEach(function (d) {\n",
    "                d[snvDataFormat.x] = +d[snvDataFormat.x];\n",
    "            });\n",
    "\n",
    "            // new lollipop chart\n",
    "            var lollipop = g3.Lollipop(\"lollipop\");\n",
    "\n",
    "            // add data\n",
    "            lollipop.data.snvData = snvData;\n",
    "            lollipop.data.domainData = domainData;\n",
    "            // add data options\n",
    "            lollipop.format.snvData = snvDataFormat;\n",
    "            lollipop.format.domainData = domainDataFormat;\n",
    "\n",
    "            // draw with default settings\n",
    "            lollipop.draw();\n",
    "\n",
    "            // get chart div id\n",
    "            var chartID = lollipop.options.chartID;\n",
    "\n",
    "            // set default filename\n",
    "            var output_chart_filename = \"lollipop\";\n",
    "\n",
    "            // button actions: download chart\n",
    "            document.getElementById(\"save-as-png\").onclick = function (e) {\n",
    "                g3.output().toPNG(output_chart_filename, chartID);\n",
    "            };\n",
    "\n",
    "            document.getElementById(\"save-as-svg\").onclick = function (e) {\n",
    "                g3.output().toSVG(output_chart_filename, chartID);\n",
    "            };\n",
    "\n",
    "        });\n",
    "    </script>\n",
    "</body>\n",
    "\n",
    "</html>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Groovy",
   "language": "groovy",
   "name": "groovy"
  },
  "language_info": {
   "codemirror_mode": "groovy",
   "file_extension": ".groovy",
   "mimetype": "",
   "name": "Groovy",
   "nbconverter_exporter": "",
   "version": "2.5.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": false,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
